From e82b368b675e391c46b7b24b9ed96bc59b6570db Mon Sep 17 00:00:00 2001
From: Lu Yahan <yahan@iscas.ac.cn>
Date: Fri, 30 Jul 2021 17:08:23 +0800
Subject: [PATCH] [riscv64][baseline] Add func FixOnHeapReferences

-- 3059689: [baseline] Fallback to handle references on heap compilation | https://chromium-review.googlesource.com/c/v8/v8/+/3059689

Change-Id: Ie055e181e5081d7acb2195ae41abaecdd0f68989
Reviewed-on: https://chromium-review.googlesource.com/c/v8/v8/+/3062030
Commit-Queue: Yahan Lu <yahan@iscas.ac.cn>
Commit-Queue: Ji Qiu <qiuji@iscas.ac.cn>
Auto-Submit: Yahan Lu <yahan@iscas.ac.cn>
Reviewed-by: Ji Qiu <qiuji@iscas.ac.cn>
Cr-Commit-Position: refs/heads/master@{#76014}
---
 src/codegen/riscv64/assembler-riscv64.cc | 2 ++
 src/codegen/riscv64/assembler-riscv64.h  | 5 +++++
 2 files changed, 7 insertions(+)

diff --git a/src/codegen/riscv64/assembler-riscv64.cc b/src/codegen/riscv64/assembler-riscv64.cc
index e83f062067..f574d0aaca 100644
--- a/src/codegen/riscv64/assembler-riscv64.cc
+++ b/src/codegen/riscv64/assembler-riscv64.cc
@@ -2758,6 +2758,8 @@ void Assembler::RelocateRelativeReference(RelocInfo::Mode rmode, Address pc,
   }
 }
 
+void Assembler::FixOnHeapReferences() { UNIMPLEMENTED(); }
+
 void Assembler::GrowBuffer() {
   DEBUG_PRINTF("GrowBuffer: %p -> ", buffer_start_);
   // Compute new buffer size.
diff --git a/src/codegen/riscv64/assembler-riscv64.h b/src/codegen/riscv64/assembler-riscv64.h
index a38428e310..0fb6414332 100644
--- a/src/codegen/riscv64/assembler-riscv64.h
+++ b/src/codegen/riscv64/assembler-riscv64.h
@@ -354,6 +354,11 @@ class V8_EXPORT_PRIVATE Assembler : public AssemblerBase {
   // ---------------------------------------------------------------------------
   // Code generation.
 
+  // This function is called when on-heap-compilation invariants are
+  // invalidated. For instance, when the assembler buffer grows or a GC happens
+  // between Code object allocation and Code object finalization.
+  void FixOnHeapReferences();
+
   // Insert the smallest number of nop instructions
   // possible to align the pc offset to a multiple
   // of m. m must be a power of 2 (>= 4).
-- 
2.35.1

